mindmeld/components/request.py

```diff
@@ -332,32 +456,84 @@ class Request:
     domain = attr.ib(default=None)
     intent = attr.ib(default=None)
     entities = attr.ib(
-        default=attr.Factory(tuple), converter=tuple_elems_to_immutable_map
+        default=attr.Factory(tuple), converter=deserialize_to_list_immutable_maps
     )
     history = attr.ib(
-        default=attr.Factory(tuple), converter=tuple_elems_to_immutable_map
+        default=attr.Factory(tuple), converter=deserialize_to_list_immutable_maps
     )
     text = attr.ib(default=None)
     frame = attr.ib(default=immutables.Map(), converter=immutables.Map)
     params = attr.ib(default=FrozenParams())
     context = attr.ib(default=immutables.Map(), converter=immutables.Map)
     confidences = attr.ib(default=immutables.Map(), converter=immutables.Map)
     nbest_transcripts_text = attr.ib(
-        default=attr.Factory(tuple), converter=tuple_elems_to_immutable_map
+        default=attr.Factory(tuple), converter=tuple
     )
     nbest_transcripts_entities = attr.ib(
-        default=attr.Factory(tuple), converter=tuple_elems_to_immutable_map
+        default=attr.Factory(tuple), converter=deserialize_to_lists_of_list_of_immutable_maps
     )
     nbest_aligned_entities = attr.ib(
-        default=attr.Factory(tuple), converter=tuple_elems_to_immutable_map
+        default=attr.Factory(tuple), converter=deserialize_to_lists_of_list_of_immutable_maps
     )
 
-    def to_dict(self):
-        return {
-            "text": self.text,
-            "domain": self.domain,
-            "intent": self.intent,
-            "context": dict(self.context),
-            "params": self.params.to_dict(),
-            "frame": dict(self.frame),
-        }
+    def to_dict(self) -> Dict[str, Any]:
+        return request_schema.dump(self)
+
+
+class RequestSchema(Schema):
+    text = fields.String(required=True)
+    domain = fields.String()
+    intent = fields.String()
+    entities = fields.Method("serialize_entities",
+                             deserialize="deserialize_list_of_maps")
+    history = fields.Method("serialize_history",
+                            deserialize="deserialize_list_of_maps")
+    params = fields.Nested(ParamsSchema)
+    frame = fields.Method("serialize_frame",
```

Comment:
Similar to the DialogueResponderSchema, an attribute for form here?

Reply:
done